packages/core/src/runtime-engine.ts

```diff
@@ -68,6 +70,19 @@ export class RuntimeEngine<ENV extends AnyEnvironment = AnyEnvironment> {
         return this;
     }
 
+    private preRun(entryDeps: FeatureClass[]) {
+        // populate required features
+        const toProcess = [...entryDeps];
+        while (toProcess.length) {
+            const feature = toProcess.shift()!;
+            if (this.allRequiredFeatures.has(feature.id)) continue;
+            this.allRequiredFeatures.add(feature.id);
+            for (const dep of feature.dependencies()) {
```

Comment:
move the check to the loop `if (this.allRequiredFeatures.has(feature.id)) continue;`
it will be much better

Reply:
because most of the features are coming from deps and not top level.